### PR TITLE
DROOLS-4495 : GDST to XLS export: Work item, Metadata and complex condition column support

### PR DESCRIPTION
**Pull request by Rikkola**
https://issues.jboss.org/browse/DROOLS-4495

### CODE REVIEW

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/DataBuilder.java [hunk 1]
```diff
@@ -26,24 +26,32 @@
...
 
+import static org.drools.workbench.models.datamodel.rule.BaseSingleFieldConstraint.TYPE_PREDICATE;
+import static org.drools.workbench.models.datamodel.rule.BaseSingleFieldConstraint.TYPE_RET_VALUE;
+import static org.drools.workbench.screens.guided.dtable.backend.server.conversion.Util.isNullOrNotNull;
+
 public class DataBuilder {
 
     private final static int FIRST_DATA_ROW = 9;
-    private final static String DROOLS_DATE_TIME_FORMAT_KEY = "drools.dateformat";
-    private final static String DEFAULT_DATE_TIME_FORMAT = "dd-MMM-yyyy";
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/DataBuilder.java]: This was dead code.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/SubHeaderBuilder.java [hunk 2]
```diff
@@ -131,26 +198,91 @@ private void makeInsert(final ActionInsertFactCol52 column) {
...
+        if (constraintValueType == TYPE_RET_VALUE) {
+            return "%s %s ( $param )";
+        } else {
+            return "%s %s $param";
+        }
+    }
+
+    private String getTemplateWithBinds(final int constraintValueType) {
+        if (constraintValueType == TYPE_RET_VALUE) {
+            return "%s : %s %s ( $param )"; // TODO test?
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/SubHeaderBuilder.java]: Need to test this.

**Comment by Rikkola** [comment]
@karreiro @jomarko if you could be kind enough to review.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/SubHeaderBuilder.java [hunk 3]
```diff
@@ -64,14 +83,14 @@ public void build() {
 
         final List<BaseColumn> expandedColumns = dtable.getExpandedColumns();
 
-        for (int sourceIndex = 0; sourceIndex < expandedColumns.size(); sourceIndex++) {
+        for (; sourceColumnIndex < expandedColumns.size(); sourceColumnIndex++) {
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/SubHeaderBuilder.java]: I am reading the class again again and I think we do not need `sourceColumnIndex` to be class variable. I think we can keep its scope just in the `for` loop. Then I think the only usage of `sourceColumnIndex` can be replaced by `columnIndex`.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/SubHeaderBuilder.java [hunk 4]
```diff
@@ -122,7 +185,11 @@ private void makeInsert(final ActionInsertFactCol52 column) {
             makeHeaderAndTitle(ACTION,
                                "");
 
-            fieldRow.createCell(columnIndex).setCellValue(column.getFactType() + " " + column.getBoundName() + " = new " + column.getFactType() + "(); insert( " + column.getBoundName() + " );");
+            fieldRow.createCell(columnIndex).setCellValue(String.format("%s %s = new %s(); insert( %s );",
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/SubHeaderBuilder.java]: Maybe also here we could use `MessageFormat` as we pass twice `column.getBoundName()`

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/XLSBuilder.java [hunk 5]
```diff
@@ -53,9 +53,12 @@ public BuildResult build() {
             makeImports();
 
             makeTable();
-        } catch (final Exception e) {
+        } catch (final UnsupportedOperationException e) {
             return new BuildResult(workbook,
                                    new XLSConversionResult(e.getMessage()));
+        } catch (final Exception e) {
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/XLSBuilder.java]: Why we added this catch block please? I do not see new `throw new` statement? Just discussing, are you aware of the place that cause this catch block is needed or is it just to be hundred percent sure?

Comment (Rikkola) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/XLSBuilder.java]: It is just to be sure. Even if this feature was complete. We still have old GDST formats that are likely to cause trouble. So this I'm hoping this picks them up and gives a cleaner report.

Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/XLSBuilder.java]: ok

**Comment by Rikkola** [comment]
@jomarko Updated. The I couldn't get the retract column to break, but might also be the changes fixed it. What I meant with complex condition column is all the different options you can set for a regular condition column.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/DataBuilder.java [hunk 6]
```diff
@@ -208,9 +261,18 @@ private String fixStringValue(final DTCellValue52 cell) {
 
         private boolean isTheRealCellValueString(final int sourceColumnIndex) {
             return !(dtable.getExpandedColumns().get(sourceColumnIndex) instanceof AttributeCol52)
+                    && !(dtable.getExpandedColumns().get(sourceColumnIndex) instanceof MetadataCol52)
+                    && !isFormula(sourceColumnIndex)
                     && Objects.equals(DataType.DataTypes.STRING, utilsWithNoRespectForLists.getTypeSafeType(dtable.getExpandedColumns().get(sourceColumnIndex)));
         }
 
+        private boolean isFormula(final int sourceColumnIndex) {
```
[the remark this answers is not in the document]
Comment (Rikkola) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/main/java/org/drools/workbench/screens/guided/dtable/backend/server/conversion/DataBuilder.java]: I changed this part a bit, hopefully easier to read.

**Comment by Rikkola** [comment]
@kiegroup/gatekeepers Can I get a merge please.

**Pull request merged**